### DISCUSSION TITLE
Stay in the current step

### DISCUSSION THREAD

**Discussion by Baoyuantop** [gilbarbara/react-joyride#844] [Q&A] [answered]
In a Step, my target is a form; When the user clicks Next, I check if the form is completed before deciding whether to go to the next step. Otherwise, it stays in the current Step.
I've done the rest of the logic, but what can I do to keep the user at the current Step in some cases when they click Next?

**Comment by gilbarbara** [comment, marked as answer]
Hey @Baoyuantop 

Are you using the `stepIndex` prop? With a controlled tour, you can decide when to move forward.
You could also use a custom tooltip and potentially hide the `Next` button until the user completes the form.

**Reply by Baoyuantop** [reply]
Thank you very much.
> hide the Next button until the user completes the form

I want the user to click Next, and then I will perform the form check.

> With a controlled tour, you can decide when to move forward

Yes, I use a controlled tour and stepIndex prop. And I am doing something like
```js
const handleJoyrideCallback = (data: CallBackProps) => {
  const { action, index, status, type } = data;
  if (action === ACTIONS.NEXT && stepIndex === 2) {
     // Performing form checks
     // if something is wrong, stay step 2
     setState({ run: true, stepIndex: 2 });
  }
}
```
But it doesn't work.

**Reply by gilbarbara** [reply]
**But it doesn't work.** isn't very helpful.
What happens?
Anyway, if the form checks fail, I don't think you should change the state.

Can you set up a simplified repro in codesandbox?

**Reply by Baoyuantop** [reply]
https://codesandbox.io/s/dreamy-kepler-z2cq12?file=/src/App.tsx

**Reply by gilbarbara** [reply]
Hey @Baoyuantop 

The next button changes the internal state even in a controlled mode, so you can't use the controlled mode for what you need to achieve.
You can use a custom Tooltip that disables the Next button until the form is valid.

**Reply by gilbarbara** [reply]
Check this example: https://codesandbox.io/s/react-joyride-with-form-4vezxs

**Reply by Baoyuantop** [reply]
Thank you very much, I will try
